mlir_graphblas/engine.py

```diff
@@ -333,24 +264,73 @@ def encoder(arg: MLIRSparseTensor) -> list:
-        (element_ctypes_input_type,) = element_ctypes_input_types
-        ctypes_input_types = [ctypes.POINTER(element_ctypes_input_type)]
+        input_c_types.append(pointer_type)  # allocated pointer (for free())
+        input_c_types.append(pointer_type)  # base pointer
+        input_c_types.append(ctypes.c_int64)  # offset from base
+        for _ in range(2 * len(tensor_type.dimensions)):  # dim sizes and strides
+            input_c_types.append(ctypes.c_int64)
 
-        def encoder(arg: Union[list, tuple]) -> list:
-            if not isinstance(arg, (list, tuple)):
+        dimensions = [dim.value for dim in tensor_type.dimensions]
+
+        def encoder(arg: np.ndarray) -> list:
+            if not isinstance(arg, np.ndarray):
                 raise TypeError(
-                    f"{repr(arg)} is expected to be an instance of {list} or {tuple}."
+                    f"{repr(arg)} is expected to be an instance of {np.ndarray.__qualname__}"
+                )
+            if not arg.dtype == np_type:
+                raise TypeError(f"{repr(arg)} is expected to have dtype {np_type}")
+            if not len(dimensions) == len(arg.shape):
+                raise ValueError(
+                    f"{repr(arg)} is expected to have rank {len(dimensions)} but has rank {len(arg.shape)}."
                 )
-            ArrayType = element_ctypes_input_type * len(arg)
-            encoded_elements = sum(map(element_encoder, arg), [])
-            array = ArrayType(*encoded_elements)
-            return [array]
+
+            encoded_args = [arg, arg, 0]
+
+            for dim_index, dim_size in enumerate(arg.shape):
+                expected_dim_size = dimensions[dim_index]
+                if (
+                    expected_dim_size is not None
+                    and arg.shape[dim_index] != expected_dim_size
+                ):
+                    raise ValueError(
+                        f"{repr(arg)} is expected to have size {expected_dim_size} in the {dim_index}th dimension but has size {arg.shape[dim_index]}."
+                    )
+                encoded_args.append(arg.shape[dim_index])
+
+            for dimension_index in range(len(arg.shape)):
+                stride = arg.strides[dimension_index] // arg.itemsize
+                encoded_args.append(stride)
+
+            return encoded_args
+
+    return input_c_types, encoder
+
+
+def input_llvm_pointer_to_ctypes(
+    mlir_type: _DIALECT_TYPES["llvm"]["LLVMPtr"],
+) -> Tuple[list, Callable]:
+    # Treat the pointer as an array (intended to represent a Python sequence).
+    element_ctypes_input_types, element_encoder = input_type_to_ctypes(
+        mlir_type.type
+    )
+    # element_ctypes_input_types has exactly one element type since
+    # a pointer type can only point to one type
+    (element_ctypes_input_type,) = element_ctypes_input_types
+    ctypes_input_types = [ctypes.POINTER(element_ctypes_input_type)]
+
+    def encoder(arg: Union[list, tuple]) -> list:
```

Comment:
I think we could also add a more specific return type hint here as well like `def encoder(arg: Union[list, tuple]) -> List[ctypes.Array]:`. I should've added this when I originally added this code. 

Again, it's not super important, so feel free to ignore this suggestion.